### DISCUSSION TITLE
Fetching global data from API at build-time?

### DISCUSSION THREAD

**Discussion by stfnsr** [vercel/next.js#15373] [Help] [answered]
Hi!
Does anyone know of a good way (or any way, really) to fetch data so that it is globally available during static generation, without explicitly adding it to getStaticProps on every single page?

My use case involves getting translations and some configuration data from an API, and my attempts didn't work out so far.

I considered doing 'something with webpack', or having a pre-build step to download the data before next kicks in, but maybe there are other options?

**Comment by CaptWeiss** [comment]
Have you considered using HOC and wrapping it around your app component in _app.js?
This way, you can pull the data in your High Order Component and pass it as context through the app component down the app tree.

**Comment by stfnsr** [comment]
Would this be different from fetching data in the app component itself?
From what I understood, only stuff fetched in getStaticProps would be available during static generation

**Comment by stfnsr** [comment, marked as answer]
Update: I went with writing a node script that runs before build and downloads the api response to a file, which then gets imported in app.js.
Seems to work well so far, although getStaticProps would be nicer ofc :)

**Comment by matepaiva** [comment]
@stfnsr I thought about doing the same as you, but it doesn't look like an elegant solution. I think `getStaticProps` would be much better, but it only work with the pages, not with the `_app`. :(

**Comment by matepaiva** [comment]
@stfnsr I found out that `getInitialProps` in `_app` does not disable the static optimization if your pages export `getStaticProps` ─ which is my case. So for me `getInitialProps` worked like a charm. 

```
Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps`
```

The warning above scared me because I thought I would lose every Static Optimization. But actually, it just removes the "automatic", it means, the optimization in the pages without `getStaticProps` and without `getServerSideProps`. This will rarely be the case for me.

So, everything good, right?

**Reply by rubenzup** [reply]
@matepaiva  do you have an example of passing the props of getInitialProps in app.js?

```
function MyApp({ Component, pageProps }: AppProps) {
    return <Component {...pageProps} />;
}
```

mine is currently like that, i dont know how to make a fetch in getInitialProps and sent it to my components

**Reply by pm0u** [reply]
Did this end up working out for you? Can't you also just force static gen if needed in this case by `export getStaticProps() {}` if absolutely needed?

**Comment by rubenzup** [comment]
Im using the example of Strapi CMS, and we do a fetch at only one endpoint, Im using getStaticProps and getStaticPaths witn ISG , but at each page, I have to do the fetch to the API and i wanted to be global

one issue i found it, my header and footer, i get it from the api, and i wanted to both of them be displayed at the 404 page, but isnt working.

I put getStaticProps at 404.tsx and add the header and footer at it ,it breaks the page

how could i fix that?
